### PR TITLE
Add warnings to cmake compilation

### PR DESCRIPTION
**Pull request by mgrojo**
Use the same set of warning flags that are used for qmake compilation.

See comments in #1718.

### CODE REVIEW

**Comment by mgrojo** [comment]
Travis is failing with these errors:
```
g++-5: error: unrecognized command line option ‘-Wnull-dereference’
g++-5: error: unrecognized command line option ‘-Wduplicated-cond’
g++-5: error: unrecognized command line option ‘-Wduplicated-branches’
```

It seems that GCC 5 doesn't have these warning flags. Why aren't they giving errors while using qmake, isn't qmake used with GCC 5 or aren't those flags actually used in the every qmake compilation?

**Comment by justinclift** [comment]
Reading through this:

* https://stackoverflow.com/questions/7827140/is-it-possible-to-get-gcc-to-ignore-unknown-w-command-line-options

Is the problem that it's giving an error here (stopping compilation) rather than just printing a warning then continuing?

**Comment by mgrojo** [comment]
@justinclift, it was giving an error and stopping compilation.

I've avoided the error by adding a condition to add those three warning flags when compiling with GCC 7.0 or greater.

File: CMakeLists.txt [hunk 1]
```diff
@@ -371,7 +371,10 @@ add_executable(${PROJECT_NAME}
 # Warnings
 if (CMAKE_COMPILER_IS_GNUCC)
     target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wextra -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion)
-    target_compile_options(${PROJECT_NAME} PRIVATE -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wuseless-cast)
+    target_compile_options(${PROJECT_NAME} PRIVATE -Wdouble-promotion -Wformat=2 -Wlogical-op -Wuseless-cast)
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 7.0)
```
Comment (justinclift) [inline CMakeLists.txt]: Clever idea. :smile:

**Comment by MKleusberg** [comment]
I guess we never tried the qmake version with GCC 5 - on Travis we only ever used cmake for building.

One suggestion: I would suggest giving this an extra config option named ```all_warning``` like for the qmake version which makes the extra warnings an opt-in thing. With all the warnings it can be pretty annoying to build plus it hides the more important warnings in a flood of style warnings :wink:

**Comment by mgrojo** [comment]
> One suggestion: I would suggest giving this an extra config option named `all_warning` like for the qmake version which makes the extra warnings an opt-in thing. With all the warnings it can be pretty annoying to build plus it hides the more important warnings in a flood of style warnings :wink:

I can't really read qmake files and didn't notice that. I'll add the same config option for cmake, then. I also thought that so many warnings were distractive, but wanted to replicate the same options as in qmake. I've never been able to compile with qmake, so don't know how is actually the output.

@tiendq have you, then, used the   `all_warning` for compiling with qmake?

**Comment by tiendq** [comment]
@mgrojo No, I haven't used qmake. I'm only familiar with CMake (to build all projects, not just this one) :)

**Comment by MKleusberg** [comment]
Thanks, @mgrojo :+1: Looks good :smile:

**Comment by mgrojo** [comment]
> @mgrojo No, I haven't used qmake. I'm only familiar with CMake (to build all projects, not just this one) :)

:confused: Then how did you get the warning in #1718? CMakeLists.txt did not have the warning flags until now. Is this a time travel paradox? :smile:

**Comment by tiendq** [comment]
@mgrojo I ran CMake, and it might only happen on macOS + Clang (as reported). Did you try it on macOS?

**Pull request merged**